libs/zedUpload/httputil/http.go

```diff
@@ -148,30 +150,112 @@ func ExecCmd(cmd, host, remoteFile, localFile string, objSize int64,
+				errorList = append(errorList, fmt.Sprintf("(attempt %d/%d):", attempt, maxRetries))
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				currentErrorList = []string{}
 			}
-			copiedSize += written
-			if written != chunkSize {
-				// Must have reached EOF
+			if ctx.Err() != nil {
+				errorList = append(errorList, fmt.Sprintf("ctx.Err: %v", ctx.Err()))
 				break
 			}
-			stats.Asize = copiedSize
-			if prgNotify != nil {
-				select {
-				case prgNotify <- stats:
-				default: //ignore we cannot write
+			if done {
+				break
+			}
+			if attempt > 0 {
+				time.Sleep(delay)
+				if delay < maxDelay {
+					delay = delay * 2
+				}
+			}
+			if canSeek && copiedSize > 0 {
+				req.Header.Set("Range", fmt.Sprintf("bytes=%d-", copiedSize))
+			}
+			resp, err := client.Do(req)
+			if err != nil {
+				currentErrorList = append(currentErrorList, fmt.Sprintf("get failed: %s", err))
+				continue
+			}
+			if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
+				currentErrorList = append(currentErrorList, fmt.Sprintf("bad response code: %d", resp.StatusCode))
+				err = resp.Body.Close()
+				if err != nil {
+					currentErrorList = append(currentErrorList, fmt.Sprintf("error close Body: %v", err))
+				}
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				//we do not want to process server errors here
+				break
+			}
+			newLastModified := resp.Header.Get("Last-Modified")
+			if lastModified != "" && newLastModified != lastModified {
+				// last modified changed, retry without Range
+				lastModified = newLastModified
+				canSeek = false
+				continue
+			}
+			if resp.StatusCode == http.StatusOK {
+				stats.BodyLength = int(resp.ContentLength)
+			}
+			// canSeek affects next retry
+			canSeek = resp.Header.Get("Accept-Ranges") == "bytes"
+			if copiedSize > 0 && (!canSeek || resp.StatusCode != http.StatusPartialContent) {
+				// restart from the begin if cannot seek or returned code not equals StatusPartialContent
+				err = local.Truncate(0)
+				if err != nil {
+					currentErrorList = append(currentErrorList, fmt.Sprintf("failed truncate file: %s", err))
+					continue
+				}
+				_, err = local.Seek(0, 0)
+				if err != nil {
+					currentErrorList = append(currentErrorList, fmt.Sprintf("failed seek file: %s", err))
+					continue
 				}
+				copiedSize = 0
 			}
+			var written int64
+			for {
+				var copyErr error
+				if written, copyErr = io.CopyN(local, resp.Body, chunkSize); copyErr != nil && copyErr != io.EOF {
+					if copyErr == io.ErrUnexpectedEOF {
+						// in case of UnexpectedEOF we must start from the beginning to keep consistency
+						canSeek = false
+					}
+					copiedSize += written
+					currentErrorList = append(currentErrorList, fmt.Sprintf("error from CopyN: %v", copyErr))
+					break
+				}
+				copiedSize += written
+				if written != chunkSize {
+					// Must have reached EOF
+					done = true
```

Comment:
Seems like if the content is exactly a multiple of chunkSize, then done will never be set.
I think the done for the whole object needs to be based on size reported in the Content-Range header per https://datatracker.ietf.org/doc/html/rfc7233#section-4.1

Reply:
I can see in docs (https://pkg.go.dev/io#CopyN) that `On return, written == n if and only if err == nil.`

Reply:
My point is that if a file is 3Mb and the server only returns 10kb up front (due to timeout), then you will set done even though you clearly didn't write 3Mb to the file. 

Furthermore, the original code has a bug (which you've inhertited) which is that if the file is exactly 1Mb in size then done never gets set.

Reply:
What am I missing?

Reply:
In case of 10kb download we will have written=10kb and it will be appended to copiedSize and we will hit another type of errors, not EOF. EOF not goes through Copy() inside of CopyN() it comes from CopyN() itself.

In case of equality we will have on the next iteration EOF (again from CopyN()) and zero written, so, it is not the issue I think.

Reply:
If the file is exactly N*1MByte then nor the original code nor you code ever sets done AFAICT.

Reply:
Ah - I didn't realize the logic depends on an extra copyN attempt which copies zero bytes at the end in that case.
Definitely makes sense to add a comment about that.

Reply:
But in the case of multiple range downloads, don't you need to set done=false for each client.Do()?

Reply:
Oh - not needed since incomplete downloads would be errors. But comments would be very helpful for the next reader.

Reply:
I put comment inside

Reply:
Thanks